libshpool/src/daemon/prompt.rs

```diff
@@ -96,15 +114,65 @@ pub fn inject_prefix(
     // hard to make the scanner work right.
     // TODO(julien): this will probably not work on mac
     let sentinel_cmd =
-        format!("\n{}=yes /proc/{}/exe daemon\n", PROMPT_SENTINEL_FLAG_VAR, std::process::id());
+        format!("\n{}=prompt /proc/{}/exe daemon\n", SENTINEL_FLAG_VAR, std::process::id());
     script.push_str(sentinel_cmd.as_str());
 
-    let mut pty_master = pty_master.is_parent().context("expected parent")?;
+    debug!("injecting prefix script '{}'", script);
     pty_master.write_all(script.as_bytes()).context("running prefix script")?;
 
     Ok(())
 }
 
+#[instrument(skip_all)]
+fn wait_for_startup(pty_master: &mut shpool_pty::fork::Master) -> anyhow::Result<()> {
+    let mut startup_sentinel_scanner = SentinelScanner::new(STARTUP_SENTINEL);
+    let startup_sentinel_cmd =
+        format!("\n{}=startup /proc/{}/exe daemon\n", SENTINEL_FLAG_VAR, std::process::id());
+
+    pty_master
+        .write_all(startup_sentinel_cmd.as_bytes())
+        .context("running startup sentinel script")?;
+
+    let mut buf: [u8; 2048] = [0; 2048];
+    loop {
+        let len = pty_master.read(&mut buf).context("reading chunk to scan for startup")?;
+        if len == 0 {
+            continue;
+        }
+        let buf = &buf[..len];
+        debug!("buf='{}'", String::from_utf8_lossy(buf));
+        for byte in buf.iter() {
+            if startup_sentinel_scanner.transition(*byte) {
+                // This might drop trailing data from the chunk we just read, but
+                // it should be fine since we are about to inject the prompt setup
+                // stuff anyway, and shell.rs will scan for the prompt setup sentinel
+                // in order to handle the smooth handoff.
+                return Ok(());
+            }
+        }
+    }
+}
+
+/// Determine the shell process running under the given pid by examining
+/// `/proc/<pid>/exe`.
+#[instrument(skip_all)]
+fn sniff_shell(pid: libc::pid_t) -> anyhow::Result<KnownShell> {
+    // TODO(julien): this won't work on mac, we need to figure something
```

Comment:
https://crates.io/crates/libproc

This might help.

Reply:
Looks like the relevant routine needs to be run as root: https://docs.rs/libproc/0.14.8/libproc/proc_pid/fn.pidpath.html#errors

Maybe https://docs.rs/libproc/0.14.8/libproc/proc_pid/fn.name.html will work though?

Reply:
I tried using name() and it worked like a charm, nice suggestion.